Source/TractUI.cpp

```diff
@@ -16,22 +16,123 @@ originY(0),
 {
-	t_tractProps *props = this->processor.getTractProps();
-	this->drawTongueControl(g, props);
-	this->drawTract(g, props);
+    t_tractProps *props = this->processor.getTractProps();
+    this->drawTongueControl(g, props);
+    this->drawTract(g, props);
+}
+
+void TractUI::timerCallback()
+{
+    repaint();
+}
+
+void TractUI::mouseDown(const juce::MouseEvent &e)
+{
+    double *pos;
+    t_tractProps *props = this->processor.getTractProps();
+    double x = e.getMouseDownX() - this->originX;
+    double y = e.getMouseDownY()- this->originY;
+    
+    pos = this->getEventPosition(props, x, y);
+    bool isNearTongue = this->isNearTongue(props, pos[0], pos[1]);
+    
+    if(isNearTongue){
+        this->isTongue = true;
+        this->setTongue(props, pos[0], pos[1]);
+    } else {
+        this->isTongue = false;
+        this->setConstriction(props, pos[0], pos[1]);
+    }
+}
+
+void TractUI::mouseDrag(const juce::MouseEvent &e)
+{
+    double *pos;
+    t_tractProps *props = this->processor.getTractProps();
+    double x = e.getMouseDownX() + e.getDistanceFromDragStartX() - this->originX;
+    double y = e.getMouseDownY() + e.getDistanceFromDragStartY() - this->originY;
+    
+    pos = this->getEventPosition(props, x, y);
+    
+    if(isTongue){
+        this->setTongue(props, pos[0], pos[1]);
+    } else {
+        this->setConstriction(props, pos[0], pos[1]);
+    }
+}
+
+void TractUI::mouseUp(const juce::MouseEvent &e)
+{
+    t_tractProps *props = this->processor.getTractProps();
+    if(!isTongue){
+        this->setConstriction(props, 0, 0);  //remove constriction
+    }
+}
+
+void TractUI::setConstriction(t_tractProps *props, double index, double diameter)
+{
+    double constrictionMin = -2.0;
+    double constrictionMax = 2.0;
+    
+    processor.constrictionX = index/props->n;
+    processor.constrictionY = (diameter - constrictionMin)/(constrictionMax - constrictionMin);
+}
+
+void TractUI::setTongue(t_tractProps *props, double index, double diameter)
+{
+    double tongueLowerIndexBound = props->bladeStart + 2;
+    double tongueUpperIndexBound = props->tipStart - 3;
+    
+    // this block can be made conditional on a flag if a 'disregard constraints' checkbox is added
+    if (index < tongueLowerIndexBound){
+        index = tongueLowerIndexBound;
+    } else if (index > tongueUpperIndexBound){
+        index = tongueUpperIndexBound;
+    }
+    if (diameter < this->innerTongueControlRadius){
+        diameter = this->innerTongueControlRadius;
+    } else if (diameter > this->outerTongueControlRadius){
+        diameter = this->outerTongueControlRadius;
+    }
+    
+    processor.tongueX = (index - tongueLowerIndexBound)/(tongueUpperIndexBound-tongueLowerIndexBound);
+    processor.tongueY = (diameter - this->innerTongueControlRadius)/(this->outerTongueControlRadius - this->innerTongueControlRadius);
+}
+
+double * TractUI::getEventPosition(t_tractProps *props, double x, double y)
+{
+    static double pos[2];
```

Comment:
This is very wrong 😆. Definitely don't abuse a static variable like this to return two values. Much more typical would be to pass arguments by reference, which can be filled in by the function. So something like:

void * TractUI::getEventPosition(t_tractProps *props, double x, double y, double &outX, double &outY)

See https://www.educba.com/c-plus-plus-pass-by-reference/